python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,333 @@
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.tune.result import NODE_IP
+from ray.util import PublicAPI
+from ray.util.annotations import DeveloperAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager.
+
+    This class is used to track checkpoints generated by trainables and trainers in
+    order to add metadata (e.g. the result, or the node where it has been created)
+    and for bookkeeping purposes.
+    """
+
+    MEMORY = "memory"
+    PERSISTENT = "persistent"
+
+    def __init__(
+        self,
```

Comment:
There's a few problem with consolidating these concepts. I've had a draft PR which did exactly this but we ran into lots of problems.
Generally, the `ray.ml.Checkpoint` should be the user-facing API, and its mostly a wrapper to convert between different storage formats. We could use it here to move between directories and data, but we wouldn't get rid of the TrackedCheckpoint anyway, so if at all we should do it in a follow-up. Even then it would be suboptimal to have Tune depend on ray.ml (and its dependencies). 

`dir_or_data` is what it says - a pointer to a directory, or the checkpoint data (for train and memory checkpoints). `result` is the reported result which is used to look up the checkpoint score attribute. I'll clarify this in the docstring

Reply:
can it be a future?

Reply:
Yes, dir_or_data can be a future